pkg/dns/dns.go

```diff
@@ -0,0 +1,417 @@
+	servicePollInterval := 1 * time.Second
+	for {
+		svc, err = kd.kubeClient.Services(kapi.NamespaceDefault).Get(kubernetesSvcName)
+		if err != nil || svc == nil {
+			glog.Infof("Ignoring error while waiting for service %v: %v. Sleeping %v before retrying.", name, err, servicePollInterval)
+			time.Sleep(servicePollInterval)
+			continue
+		}
+		break
+	}
+	return
+}
+
+func (kd *KubeDNS) GetCacheAsJSON() string {
+	json, _ := kd.cache.Serialize("")
+	return json
+}
+
+func (kd *KubeDNS) setServicesStore() {
+	// Returns a cache.ListWatch that gets all changes to services.
+	serviceWatch := kcache.NewListWatchFromClient(kd.kubeClient, "services", kapi.NamespaceAll, kselector.Everything())
+	kd.servicesStore, kd.serviceController = kframework.NewInformer(
+		serviceWatch,
+		&kapi.Service{},
+		resyncPeriod,
+		kframework.ResourceEventHandlerFuncs{
+			AddFunc:    kd.newService,
+			DeleteFunc: kd.removeService,
+			UpdateFunc: kd.updateService,
+		},
+	)
+}
+
+func (kd *KubeDNS) setEndpointsStore() {
+	// Returns a cache.ListWatch that gets all changes to endpoints.
+	endpointsWatch := kcache.NewListWatchFromClient(kd.kubeClient, "endpoints", kapi.NamespaceAll, kselector.Everything())
+	kd.endpointsStore, kd.eController = kframework.NewInformer(
+		endpointsWatch,
+		&kapi.Endpoints{},
+		resyncPeriod,
+		kframework.ResourceEventHandlerFuncs{
+			AddFunc: kd.handleEndpointAdd,
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				// TODO: Avoid unwanted updates.
+				kd.handleEndpointAdd(newObj)
+			},
+		},
+	)
+}
+
+func (kd *KubeDNS) newService(obj interface{}) {
+	if service, ok := obj.(*kapi.Service); ok {
+		// if ClusterIP is not set, a DNS entry should not be created
+		if !kapi.IsServiceIPSet(service) {
+			kd.newHeadlessService(service)
+			return
+		}
+		if len(service.Spec.Ports) == 0 {
+			glog.Info("Unexpected service with no ports, this should not have happend: %v", service)
+		}
+		kd.newPortalService(service)
+	}
+}
+
+func (kd *KubeDNS) removeService(obj interface{}) {
+	if s, ok := obj.(*kapi.Service); ok {
+		subCachePath := append(kd.domainPath, serviceSubdomain, s.Namespace, s.Name)
+		kd.cache.DeletePath(subCachePath...)
+	}
+}
+
+func (kd *KubeDNS) updateService(oldObj, newObj interface{}) {
+	kd.newService(newObj)
+}
+
+func (kd *KubeDNS) handleEndpointAdd(obj interface{}) {
+	if e, ok := obj.(*kapi.Endpoints); ok {
+		kd.addDNSUsingEndpoints(e)
+	}
+}
+
+func (kd *KubeDNS) addDNSUsingEndpoints(e *kapi.Endpoints) error {
+	svc, err := kd.getServiceFromEndpoints(e)
+	if err != nil {
+		return err
+	}
+	if svc == nil || kapi.IsServiceIPSet(svc) {
+		// No headless service found corresponding to endpoints object.
+		return nil
+	}
+	return kd.generateRecordsForHeadlessService(e, svc)
+}
+
+func (kd *KubeDNS) getServiceFromEndpoints(e *kapi.Endpoints) (*kapi.Service, error) {
+	key, err := kcache.MetaNamespaceKeyFunc(e)
+	if err != nil {
+		return nil, err
+	}
+	obj, exists, err := kd.servicesStore.GetByKey(key)
```

Comment:
`kcache.MetaNamespaceKeyFunc(e)` constructs the key for the endpoints object and we give that to the service's `GetByKey()`. This only works because corresponding endpoints and the service objects have identical names. I think it is worth adding a comment about it here.